File: UCB1/UCB1.cpp
/*************************************************************************
    > File Name: UCB1.cpp
    > Author: YinWen
    > Mail: YinWenatBIT@163.com 
    > Created Time: Tue 01 Sep 2015 08:32:35 PM CST
 ************************************************************************/

#include "UCB1.h"
#include <cmath>
#include <stdio.h>
#include <cstdlib>
#include "json/json.h"
#include <array>



using namespace std;

UCB1::UCB1():
    totalcount(0), default_count(0), default_value(0.0)
{

}

UCB1::UCB1(int init_totalcount, int init_count, double init_value):
	totalcount(init_totalcount), default_count(init_count), default_value(init_value) 
{

}

UCB1::~UCB1()
{

}

string UCB1::toString()
{
    Json::FastWriter writer;
    Json::Value value;
    Json::Value element;
    std::string key;

    value["totalcount"] = Json::Value(totalcount);
    value["default_count"] = Json::Value(default_count);
    value["default_value"] = Json::Value(default_value);

    for(auto it = frequencyReward.begin(); it !=frequencyReward.end(); it++)
    {
        key = it->first;

        Json::Value temp_value;
        temp_value["counts"] = Json::Value(it->second.counts);
        temp_value["keyValue"] = Json::Value(it->second.values);
        element[key] = temp_value;
    }
    value["element"] = element;
    return writer.write(value);
}


bool UCB1::readFromString(const string& JString)
{
    Json::Reader reader;
    Json::Value value;
    Json::Value element;
    std::string key;

    int readcount;
    double readdouble;

    if(!reader.parse(JString, value))
        return false;

    element = value["element"];
    auto keymember = element.getMemberNames();
    int num = keymember.size();

    for(int i=0; i<num; i++)
    {
        key = keymember[i];
        readcount = element[key]["counts"].asInt();
        readdouble = element[key]["keyValue"].asDouble();

        frequencyReward.insert({key, {readcount, readdouble}});
    }

    totalcount = value["totalcount"].asInt();
    default_count = value["default_count"].asInt();
    default_value = value["default_value"].asDouble();
    return true;
}

bool UCB1::insert(string &key)
{
    auto ret = frequencyReward.insert({key, {default_count, default_value}});
    return ret.second;
}


bool UCB1::insert(string& key, UCB value)
{
    auto ret = frequencyReward.insert({key, value});
    return ret.second;
}

void UCB1::add_totalcount(int num)
{
	totalcount += num;
}

void UCB1::set_totalcount(int number)
{
	totalcount = number;
}


string UCB1::get_totalcount()
{
	char temp[25];
	sprintf(temp,"%d", totalcount);
	string totalcountstr(temp);
	return totalcountstr;
}

string UCB1::get_count(string key)
{
	string countstr;
	auto origin = frequencyReward.find(key);
	if(origin == frequencyReward.end())
		return countstr;

	char temp[25];
	sprintf(temp, "%d", origin->second.counts);
	countstr = temp;
	return countstr;
}

string UCB1::get_value(string key)
{
	string valuestr;
	auto origin = frequencyReward.find(key);
	if(origin == frequencyReward.end())
		return valuestr;

	char temp[25];
	sprintf(temp, "%f", origin->second.values);
	valuestr = temp;
	return valuestr;
}


bool UCB1::update(const char * start, double res)
{
    string key(start);
    return update(key, res);
}



bool UCB1::update(string& key, double res)
{
	auto origin = frequencyReward.find(key);
	if(origin == frequencyReward.end())
    {
	    auto ret = frequencyReward.insert({ key, {default_count, default_value} });
        return true;
    }


	double n = ++origin->second.counts;
	origin->second.values = origin->second.values*(n-1)/n + res/n;
//    ++this->totalcount;

	return true;
}

bool UCB1::update_reset_last(string & key, double last, double res)
{
	auto origin = frequencyReward.find(key);
	if(origin == frequencyReward.end())
    {
	    auto ret = frequencyReward.insert({ key, {default_count, default_value} });
        return true;
    }


	double n = origin->second.counts;
	origin->second.values = (origin->second.values*n -last + res)/n;

	return true;

}
bool UCB1::update_reset_last(const char *start, double last, double res)
{
    string key(start);
    return update_reset_last(key, last, res);
}

string UCB1::select_arm()
{
	string maxkey;

	double bonus;
	double maxvalue = 0.0;
   
    if(frequencyReward.empty())
        throw empty_arm("no arm in the map");

	for(auto it = frequencyReward.begin(); it != frequencyReward.end(); it++ )
	{
		if(it->second.counts == 0)
			return it->first;
		
		bonus = sqrt(2* log((double)totalcount))/it->second.counts;
		if(maxvalue < bonus + it->second.values)
		{
			maxvalue = bonus + it->second.values;
			maxkey = it->first;
		}
	}

	return maxkey;
}


std::vector<std::string> & UCB1::select_arm_N(size_t n)
{
    keystrs.clear();
    int countzero= 0;

    if(n ==0)
        return keystrs;
    if(n > frequencyReward.size())
        n = frequencyReward.size();

    if(frequencyReward.empty())
        throw empty_arm("no arm in the map");



    vector<std::string> maxkey;
    maxkey.resize(n+1);

	double bonus;
    double valuenow;
    vector<double> maxvalue;
    maxvalue.resize(n+1, 0.0);

	for(auto it = frequencyReward.begin(); it != frequencyReward.end() && countzero <n ; it++ )
	{
		if(it->second.counts == 0)
        {
			keystrs.push_back(it->first);
            ++countzero;
        }
        else
        {		
	    	bonus = sqrt(2* log((double)totalcount))/it->second.counts;
            valuenow = bonus +it->second.values;

            int i=n;
            while(i>0)
            {
                if(valuenow >= maxvalue[i-1])
                {
                    maxvalue[i] = maxvalue[i-1];
                    maxkey[i] = maxkey[i-1];
                    --i;   
                }
                else
                    break;

            }
            maxvalue[i] = valuenow;
            maxkey[i] = it->first;


    	}
    }

    int remain = n - countzero;
    int i=0;
    while(i< remain)
    {
        keystrs.push_back(maxkey[i]);
        i++;
    }

    for(int i=0; i<keystrs.size(); i++)
        update(keystrs[i], default_count);

    this->add_totalcount(1);

    return keystrs;


}

File: UCBAlgorithm/UCBServer_python/UCB1.py
#!/usr/bin/env python
# coding=utf-8

import math
import simplejson
__metaclass__  = type

class UCB1:
    def __init__(self, init_totalcount = 0, init_count =0, init_value =0.0):
        self.totalcount = init_totalcount
        self.default_count = init_count
        self.default_value = init_value
        self.frequencyReward = {}

    def get_totalcount(self):
        return self.totalcount

    def get_count(self, keystr):
        try:
            return self.frequencyReward[keystr][0]
        except KeyError:
            return None

    def get_value(self, keystr):
        try:
            return self.frequencyReward[keystr][1]
        except KeyError:
            return None

    def insert(self, key, count = None, value = None):
        if key not in self.frequencyReward:
            if count is not None and value is not None:
                self.frequencyReward[key] = [count, value, 0]
            else:
                self.frequencyReward[key] = [self.default_count, self.default_value, 0]
        else:
            pass

    def addBlackList(self, blacklist):
        for key in blacklist:
            self.frequencyReward[key][2] =1

    def set_totalcount(self, num):
        self.totalcount = num

    
    def add_totalcount(self, num):
        self.totalcount += num

    def update(self, key, value):
        if key in self.frequencyReward:
            self.frequencyReward[key][0] += 1
            n = self.frequencyReward[key][0]
            prevalue = self.frequencyReward[key][1]
            self.frequencyReward[key][1] = (prevalue*(n-1) + value)/n
        else:
            self.insert(key)

    def update_reset_last(self, key, value):
            prevalue = self.frequencyReward[key][1]
            n = self.frequencyReward[key][0]
            self.frequencyReward[key][1] = (prevalue*n -self.default_value +value)/n


    def get_all(self):
        return self.frequencyReward

    def select_arm(self):
        maxkey = ''
        maxvalue = 0.0
        for key in self.frequencyReward:
            if self.frequencyReward[key][2] == 1:
                continue

            if self.frequencyReward[key][1] == 0.0 :
                self.add_totalcount(1)
                self.update(key, self.default_value)
                return key

            totalvalue = self.value_bonus(key)
            if totalvalue>maxvalue:
                maxvalue = totalvalue
                maxkey = key
            
        self.add_totalcount(1)
        self.update(maxkey, self.default_value)
        return maxkey
    
    #添加黑名单后，立刻重新选择新的推荐，但是不增加totalcount与counts
    def select_arm_N_forcedbyBlackList(self, N):
        if N == 0:
            return None
        
        allvalue = {x:self.value_bonus(x) for x in self.frequencyReward if self.frequencyReward[x][2] != 1}

        sorted_key = sorted(allvalue.iteritems(), key = lambda asd:asd[1], reverse = True)
        allstr = [x[0] for x in sorted_key]
        return allstr[0:N]


    def select_arm_N(self, N):
        if N == 0:
            return None
        
        allvalue = {x:self.value_bonus(x) for x in self.frequencyReward if self.frequencyReward[x][2] != 1}

        sorted_key = sorted(allvalue.iteritems(), key = lambda asd:asd[1], reverse = True)
        allstr = [x[0] for x in sorted_key]
        for i in range(N):
            self.update(allstr[i], self.default_value)

        self.add_totalcount(1)
        return allstr[0:N]
        

    def value_bonus(self, key):
        if key in self.frequencyReward:
            bonus = math.sqrt(2*math.log(self.totalcount)) /float(self.frequencyReward[key][0])
            return bonus+self.frequencyReward[key][1]


    def select_arm_N_sort(self, N):
        if N == 0:
            return None
        
        allvalue = {x:self.value_bonus(x) for x in self.frequencyReward}
        sorted(allvalue.iteritems(), key = lambda asd:asd[1], reverse = True)
        return allvalue

    def toString(self):
        json = {}
        json["totalcount"] = self.totalcount
        json["default_count"] = self.default_count
        json["default_value"] = self.default_value
        json["element"] = self.frequencyReward
        #element = simplejson.dumps(self.frequencyReward)
        json_str = simplejson.dumps(json)
        return json_str

    def readFromString(self, str):
        try:
            readDict = simplejson.loads(str)
            self.totalcount = readDict["totalcount"]
            self.default_count = readDict["default_count"]
            self.default_value = readDict["default_value"]
            self.frequencyReward = readDict["element"]
        except :
            print "wrong string"
